num_parse: support years (nineteen twenty)

dimidd opened the pull request:
Signed-off-by: Dimid Duchovny <dimidd@gmail.com>

markburns commented:
Hey sorry I somehow didn't get a notification of this, so only just seen it.

I'm not sure if I agree this should go in the gem. It seems like it would open the door up to lots of invalid values being parseable, even the test values you put look wrong. 

Maybe it can be added as a separate helper method, and only then if it supports two-pair values. I can't see a valid usecase for `"forty nine ninety eight forty seven seventy nine".in_numbers .should == 49984779`

Also, an aside, but I'm not going to merge if you change the contact email and repo over to your name.
(I can understand adding an author if there is a significant contribution)

dimidd commented:
>Hey sorry I somehow didn't get a notification of this, so only just seen it.
>

Hi Mark

>I'm not sure if I agree this should go in the gem. It seems like it would open the door up to lots of invalid values being parseable, even the test values you put look wrong.
>
>Maybe it can be added as a separate helper method, and only then if it supports two-pair values. I can't see a valid usecase for "forty nine ninety eight forty seven seventy nine".in_numbers .should == 49984779
>

I use it as a part of a speech-to-text analyzer, where people dictate numbers with many digits after the decimal point.

>Also, an aside, but I'm not going to merge if you change the contact email and repo over to your name.
>(I can understand adding an author if there is a significant contribution)

Sorry, since the last commit was more than 2 years ago, and I my PR wasn't replied, I had the impression the gem isn't actively maintained. So I took the liberty to change the contact details. I'll revert now.

>This diff is a bit hard to mentally parse. Could you explain what this fixes, or better, add a failing test that this fixed?

I don't remember the specific case, but I realized the `compress` method was needlessly complex, so I rewrote it. I'll squash it to the original commit.

markburns commented:
Nice! That is actually a valid use-case. I hadn't considered something like
that.

Based on that I'll want to merge something in, but it would be nice if it
was just a separate method I think. Not sure of a good name.
On Tue, Apr 14, 2015 at 5:42 PM dimidd <notifications@github.com> wrote:

> Hey sorry I somehow didn't get a notification of this, so only just seen
> it.
>
> Hi Mark
>
> I'm not sure if I agree this should go in the gem. It seems like it would
> open the door up to lots of invalid values being parseable, even the test
> values you put look wrong.
>
> Maybe it can be added as a separate helper method, and only then if it
> supports two-pair values. I can't see a valid usecase for "forty nine
> ninety eight forty seven seventy nine".in_numbers .should == 49984779
>
> I use it as a part of a speech-to-text analyzer, where people dictate
> numbers with many digits after the decimal point.
>
> Also, an aside, but I'm not going to merge if you change the contact email
> and repo over to your name.
> (I can understand adding an author if there is a significant contribution)
>
> Sorry, since the last commit was more than 2 years ago, and I my PR wasn't
> replied, I had the impression the gem isn't actively maintained. So I took
> the liberty to change the contact details. I'll revert now.
>
> This diff is a bit hard to mentally parse. Could you explain what this
> fixes, or better, add a failing test that this fixed?
>
> I don't remember the specific case, but I realized the compress method
> was needlessly complex, so I rewrote it. I'll squash it to the original
> commit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/markburns/numbers_in_words/pull/3#issuecomment-92697450>
> .
>

markburns commented:
I just pulled this into my local repo to try and understand it. 

Thanks for putting in the work on this.

Right now, I won't be able to merge this as it is because of e.g. https://github.com/dimidd/numbers_in_words/blob/master/lib/numbers_in_words/to_number.rb#L36

Hard-coding English terms there and not in the language specific class is a bit of an anti-pattern. It opposes the principle of separating out the language specific stuff from the number parsing stuff.

It would be nice we did use the values from the language specific classes.

dimidd commented:
I see, thanks. Should I delegate `handle_negative, handle_decimals, strip_punctuation` to `:language`?

markburns commented on lib/numbers_in_words/number_parser.rb:
```diff
@@ -39,7 +39,14 @@ module NumbersInWords::NumberParser
+    if integers.length < 2
+      return only_compress ? integers : integers.empty? ? 0 : integers[0]
+    end
+    if [] == scales_n & integers
```
Surely this could never be true? You just assigned it on line 43

markburns replied:
Just looked at my comment from before and I said the same thing in April. This to me is evidence enough that this is a bit confusing. Can you change it to

```ruby
  if (integers & scales_n)==[]
```

Less chance for mistaking for ([] == scales_n) && integers

dimidd replied:
How about
`[] == (scales_n & integers)` ?
I like to put the constant before the variable in comparisons, in order to make it look more distinct than assignment.

markburns replied:
Sure, I kind of prefer the other, but I don't mind

markburns commented on lib/numbers_in_words/to_number.rb:
```diff
@@ -16,27 +16,39 @@ def language
-  def handle_negative text
-    -1 * (text.gsub(/^minus /, "")).in_numbers if text =~ /^minus /
+  def handle_negative(text, only_compress)
+    if text =~ /^minus/
```
The minus needs to be delegated to the language too

markburns commented on lib/numbers_in_words/to_number.rb:
```diff
@@ -16,27 +16,39 @@ def language
+    i = handle_negative(text, only_compress)
     return i if i
 
+    mixed = text.match /^(-?\d+(.\d+)?) (hundred|thousand|million|billion|trillion)$/
```
These should come from the constant that defines them already (and is more comprehensive)

markburns commented on lib/numbers_in_words/to_number.rb:
```diff
@@ -16,27 +16,39 @@ def language
+    mixed = text.match /^(-?\d+(.\d+)?) (hundred|thousand|million|billion|trillion)$/
+    return mixed[1].in_numbers * mixed[3].in_numbers if mixed && mixed[1] && mixed[3]
+
+    one = text.match /^one (hundred|thousand|million|billion|trillion)$/
```
same here

markburns commented on numbers_in_words.gemspec:
```diff
@@ -9,13 +9,14 @@ Gem::Specification.new do |gem|
-  gem.email         = ["markthedeveloper@gmail.com"]
-  gem.homepage      = "http://github.com/markburns/numbers_in_words"
+  gem.authors       = ["Mark Burns", "Dimid Duchovny"]
+  gem.email         = ["dimidd@gmail.com"]
```
You can add your email, just don't replace mine, also leave the homepage as it is

markburns commented:
Overall looking good, if you could just address the few comments it should be ok. I'm going to add travis and codeclimate, then run rubocop, so we might have a few tidy up issues to handle but aside from that I'm happy with how it's starting to look.

markburns commented:
As for your question, not delegate the method entirely, just the language specific stuff. E.g. the multiplication would still occur inside the handle negative method, but the word "minus" would come from the language class.

Do you speak a second language? If so it might help to understand by having a go at implementing another language. I could have a go at Japanese. Then we'd be able to see any pain points and duplication.

markburns commented on lib/numbers_in_words/to_number.rb:
```diff
@@ -48,25 +60,14 @@ def handle_decimals text
     match = text.match(/\spoint\s/)
```
This match also needs to move into the language

markburns commented:
OK modernized the repo a bit, you'll need to fix the merge conflicts and we'll see rubocop issues/test coverage/code metrics scores etc when you next push to this branch.

markburns commented:
replaced by #5

markburns commented:
added you as a collaborator @dimidd - so if you get the GPA back to 4.0 and address the issues mentioned here, then we can merge it in

dimidd commented:
I added a new branch: `years`, where I fixed the issues you raised, please tell me if I missed anything.
There's one spec failing:
```Failure/Error: expect(NumbersInWords.in_words(100*10**100)).to eq "one hundred googols"```
What do you mean by GPA?

I can add more languages, but it'll take some time.
